File: architecture-skills/object-oriented-programming-paradigm/examples/object-relationships/object-relationships.swift
// Object relationships in Swift.
//
// ARC makes the distinction semi-visible: a let reference held only by its
// whole behaves as composition (deinit fires when the whole goes), strong
// references shared across objects are aggregation, and passing a collaborator
// as a parameter is association because it adds no lifetime coupling.

import Foundation

struct Book {
    let title: String
}

final class Library {
    private let books: [Book]

    init(books: [Book]) {
        self.books = books
    }

    func lookup(_ title: String) -> Book? {
        books.first { $0.title == title }
    }
}

final class Professor {
    let name: String

    init(name: String) {
        self.name = name
    }

    // Association: library is a parameter — no stored reference.
    func lookupBook(_ library: Library, title: String) -> String {
        guard let book = library.lookup(title) else {
            return "\(name): '\(title)' not in catalog"
        }
        return "\(name) looked up '\(book.title)' in the library"
    }
}

struct Course {
    let title: String
}

final class Department {
    let name: String
    private var professors: [Professor] // aggregation: strong refs, shared
    private var courses: [Course] = []  // composition: owned exclusively here

    init(name: String, professors: [Professor]) {
        self.name = name
        self.professors = professors
    }

    func addCourse(_ title: String) {
        courses.append(Course(title: title))
    }

    func dissolve() {
        professors.removeAll() // drop strong refs — professors survive if held elsewhere
        courses.removeAll()    // composed parts released when whole clears them
    }

    func printRoster() {
        let names = professors.map(\.name).joined(separator: ", ")
        print("\(name) roster: \(names)")
    }

    func printCourses() {
        if courses.isEmpty {
            print("\(name) courses: (none)")
            return
        }
        let titles = courses.map(\.title).joined(separator: ", ")
        print("\(name) courses: \(titles)")
    }
}

let library = Library(books: [
    Book(title: "Algorithms"),
    Book(title: "Calculus"),
])

let alice = Professor(name: "Dr. Alice")
let bob = Professor(name: "Dr. Bob")

let cs = Department(name: "Computer Science", professors: [alice, bob])
cs.addCourse("Algorithms")
cs.addCourse("Data Structures")

let math = Department(name: "Mathematics", professors: [alice])
math.addCourse("Calculus")

print("=== Association (uses-a) ===")
print(alice.lookupBook(library, title: "Algorithms"))

print("\n=== Aggregation (has-a) ===")
cs.printRoster()
math.printRoster()
cs.dissolve()
print("After dissolving CS — Alice still exists: \(alice.name)")
math.printRoster()

print("\n=== Composition (owns-a) ===")
math.printCourses()
math.dissolve()
math.printCourses()

File: architecture-skills/object-oriented-programming-paradigm/examples/object-relationships/object-relationships.cpp
// Object relationships in C++.
//
// Ownership is spelled in the member type: unique_ptr or by-value member is
// composition (destructor cascades), shared_ptr is aggregation, raw pointer
// or reference is association (non-owning by convention).

#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct Book {
    std::string title;
};

struct Library {
    std::vector<Book> books;

    const Book *lookup(const std::string &title) const {
        for (const Book &book : books) {
            if (book.title == title) {
                return &book;
            }
        }
        return nullptr;
    }
};

struct Professor {
    std::string name;

    // Association: library is a non-owning parameter.
    std::string lookupBook(const Library &library, const std::string &title) const {
        const Book *book = library.lookup(title);
        if (!book) {
            return name + ": '" + title + "' not in catalog";
        }
        return name + " looked up '" + book->title + "' in the library";
    }
};

struct Course {
    std::string title;
};

class Department {
public:
    Department(std::string name, std::vector<Professor *> professors)
        : name_(std::move(name)), professors_(std::move(professors)) {}

    void addCourse(std::string title) {
        courses_.push_back(std::make_unique<Course>(Course{std::move(title)}));
    }

    std::vector<std::string> professorNames() const {
        std::vector<std::string> names;
        for (const Professor *p : professors_) {
            names.push_back(p->name);
        }
        return names;
    }

    std::vector<std::string> courseTitles() const {
        std::vector<std::string> titles;
        for (const auto &c : courses_) {
            titles.push_back(c->title);
        }
        return titles;
    }

    void dissolve() {
        professors_.clear();  // aggregation: never delete professors
        courses_.clear();     // composition: unique_ptrs destroyed here
    }

    void printRoster() const {
        std::cout << name_ << " roster:";
        for (const Professor *p : professors_) {
            std::cout << " " << p->name;
        }
        std::cout << "\n";
    }

    void printCourses() const {
        if (courses_.empty()) {
            std::cout << name_ << " courses: (none)\n";
            return;
        }
        std::cout << name_ << " courses:";
        for (const auto &c : courses_) {
            std::cout << " " << c->title;
        }
        std::cout << "\n";
    }

private:
    std::string name_;
    std::vector<Professor *> professors_;              // aggregation: raw ptr
    std::vector<std::unique_ptr<Course>> courses_;     // composition: sole owner
};

int main() {
    Library library{{{"Algorithms"}, {"Calculus"}}};

    Professor alice{"Dr. Alice"};
    Professor bob{"Dr. Bob"};

    Department cs("Computer Science", {&alice, &bob});
    cs.addCourse("Algorithms");
    cs.addCourse("Data Structures");

    Department math("Mathematics", {&alice});
    math.addCourse("Calculus");

    std::cout << "=== Association (uses-a) ===\n";
    std::cout << alice.lookupBook(library, "Algorithms") << "\n";

    std::cout << "\n=== Aggregation (has-a) ===\n";
    cs.printRoster();
    math.printRoster();
    cs.dissolve();
    std::cout << "After dissolving CS — Alice still exists: " << alice.name << "\n";
    math.printRoster();

    std::cout << "\n=== Composition (owns-a) ===\n";
    math.printCourses();
    math.dissolve();
    math.printCourses();

    return 0;
}
